### DISCUSSION TITLE
Cannot get started contributing on MacOS

### DISCUSSION THREAD

**Discussion by iamkirkbater** [containers/podman#16066] [General]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

I guess it could be either, depending on which way you look at it.

**Description**

I'm running MacOS and want to contribute to Podman. I'm having an issue trying to even build the podman CLI and can't help contribute.

**Steps to reproduce the issue:**

1. Run MacOS
2. All documentation on getting started is geared towards Linux, so I'm not sure what prereqs are relevant, but I've made a best-effort attempt (although my best effort is probably not great)
3. run `make binaries`:

**Describe the results you received:**
```> make binaries
Makefile:134: invalid `override' directive
Podman is being compiled without the systemd build tag.         Install libsystemd on Ubuntu or systemd-devel on rpm based      distro for journald support.
CGO_ENABLED=0 GOOS=darwin GOARCH=arm64 go build \
                 \
                -ldflags '-X github.com/containers/podman/v4/libpod/define.gitCommit=6826bf427069ae2e8ac212802e8b60a3f433867c -X github.com/containers/podman/v4/libpod/define.buildInfo=1664995096 -X github.com/containers/podman/v4/libpod/config._installPrefix=/usr/local -X github.com/containers/podman/v4/libpod/config._etcDir=/usr/local/etc -X github.com/containers/common/pkg/config.additionalHelperBinariesDir= ' \
                -tags " exclude_graphdriver_btrfs btrfs_noversion    exclude_graphdriver_devicemapper seccomp" \
                -o bin/podman ./cmd/podman
package github.com/containers/podman/v4/cmd/podman
        imports github.com/containers/podman/v4/cmd/podman/common
        imports github.com/containers/podman/v4/cmd/podman/registry
        imports github.com/containers/podman/v4/pkg/domain/infra
        imports github.com/containers/podman/v4/libpod
        imports github.com/containers/common/libnetwork/cni: build constraints exclude all Go files in /Users/kbater/Projects/containers/podman/vendor/github.com/containers/common/libnetwork/cni
package github.com/containers/podman/v4/cmd/podman
        imports github.com/containers/podman/v4/cmd/podman/common
        imports github.com/containers/podman/v4/cmd/podman/registry
        imports github.com/containers/podman/v4/pkg/domain/infra
        imports github.com/containers/podman/v4/libpod
        imports github.com/containers/common/libnetwork/network: build constraints exclude all Go files in /Users/kbater/Projects/containers/podman/vendor/github.com/containers/common/libnetwork/network
make: *** [bin/podman] Error 1```

**Describe the results you expected:**
go builds the podman binary into `bin/podman`

**Additional information you deem important (e.g. issue happens only occasionally):**
Running this on MacOS Monterrey, 12.6, M1Pro 32GB RAM

**Output of `podman version`:**

```
> podman version
Client:       Podman Engine
Version:      4.2.1
API Version:  4.2.1
Go Version:   go1.18.6
Built:        Tue Sep  6 15:16:02 2022
OS/Arch:      darwin/arm64

Server:       Podman Engine
Version:      4.2.1
API Version:  4.2.1
Go Version:   go1.18.5
Built:        Wed Sep  7 15:59:25 2022
OS/Arch:      linux/arm64
```

**Output of `podman info`:**

```
host:
  arch: arm64
  buildahVersion: 1.27.0
  cgroupControllers:
  - cpu
  - io
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.4-2.fc36.aarch64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.4, commit: '
  cpuUtilization:
    idlePercent: 99.06
    systemPercent: 0.72
    userPercent: 0.22
  cpus: 4
  distribution:
    distribution: fedora
    variant: coreos
    version: "36"
  eventLogger: journald
  hostname: localhost.localdomain
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 1000000
    uidmap:
    - container_id: 0
      host_id: 501
      size: 1
    - container_id: 1
      host_id: 100000
      size: 1000000
  kernel: 5.19.12-200.fc36.aarch64
  linkmode: dynamic
  logDriver: journald
  memFree: 5393612800
  memTotal: 6143082496
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.6-2.fc36.aarch64
    path: /usr/bin/crun
    version: |-
      crun version 1.6
      commit: 18cf2efbb8feb2b2f20e316520e0fd0b6c41ef4d
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/501/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-0.2.beta.0.fc36.aarch64
    version: |-
      slirp4netns version 1.2.0-beta.0
      commit: 477db14a24ff1a3de3a705e51ca2c4c1fe3dda64
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 0
  swapTotal: 0
  uptime: 0h 12m 37.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /var/home/core/.config/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/home/core/.local/share/containers/storage
  graphRootAllocated: 63876083712
  graphRootUsed: 2001530880
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 0
  runRoot: /run/user/501/containers
  volumePath: /var/home/core/.local/share/containers/storage/volumes
version:
  APIVersion: 4.2.1
  Built: 1662580765
  BuiltTime: Wed Sep  7 15:59:25 2022
  GitCommit: ""
  GoVersion: go1.18.5
  Os: linux
  OsArch: linux/arm64
  Version: 4.2.1

```

**Package info (e.g. output of `rpm -q podman` or `apt list podman` or `brew info podman`):**

```
==> podman: stable 4.2.1 (bottled), HEAD
Tool for managing OCI containers and pods
https://podman.io/
/opt/homebrew/Cellar/podman/4.2.1 (178 files, 48MB) *
  Poured from bottle on 2022-09-27 at 12:02:00
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/podman.rb
License: Apache-2.0
==> Dependencies
Build: go-md2man ✘, go@1.18 ✘
Required: qemu ✔
==> Options
--HEAD
        Install HEAD version
==> Caveats
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions
==> Analytics
install: 24,813 (30 days), 64,356 (90 days), 211,186 (365 days)
install-on-request: 24,012 (30 days), 62,808 (90 days), 209,311 (365 days)
build-error: 1 (30 days)
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical

I'd love to know where I can help get started with the project, I'm eager to help bring this to MacOS.

**Comment by baude** [comment]
hello @iamkirkbater ... you are correct; most things are geared towards starting on Linux and contributing from Linux.  But fear not, there are options!

the first thing to know is that on darwin, try make podman-remote.  this will build the native macos "remote" client, which is the client for Macs.  In this case, consider the client to be the "front" end.  The "backend" runs on Linux and only on Linux.  So if this is the case, you will need to create a VM of your choice and run Linux there to interact with it.

we can probably be more realtime about helping you if you are able to join us on irc -> https://podman.io/community/#slack-irc-matrix-and-discord

I'm going to convert this to a discussion as it is a perfect fit for that.

**Reply by iamkirkbater** [reply]
That makes so much more sense!  Thanks for this pointer, that command worked very well.

Of course I start trying to look through the code to figure out where the docs are to start writing this down, and then I find https://github.com/containers/podman/blob/main/build_osx.md 🤦🏼 

I'll have a PR up by the end of the day to add a link to that page from the CONTRIBUTING.md page where I got lost before.

Thanks so much for your time!
